state_machine/applications/system-check/tests/imu_test.py

```diff
@@ -0,0 +1,193 @@
+
+    elif test_type == "Rotating":
+        # print user prompts
+        print("Please rotate the cubesat as best as possible. Waiting",
+              wait_time, "seconds.")
+        time.sleep(wait_time)
+        print("Collecting IMU data...")
+
+        # at some random time, collect acc and gyro readings, return 0.0 mag
+        random_wait_time = random.randint(1, wait_time)
+        time.sleep(random_wait_time)
+        acc = cubesat.acceleration()
+        gyro = cubesat.gyro()
+        time.sleep(wait_time - random_wait_time)
+        print("Data Collection Complete")
+        return acc, gyro, 0.0
+
+    elif test_type == "Magnet":
+        # print user prompts
+        print("Please leave the cubesat flat on a table and retrieve a " +
+              "magnet. " + "Waiting", wait_time, "seconds")
+        time.sleep(wait_time)
+        print("Please slowly move the magnet closer to the cubesat for",
+              wait_time, "seconds")
+        print("Collecting IMU data...")
+
+        # collect acc, gyro, mag at the start, and mag at end
+        acc = cubesat.acceleration()
+        gyro = cubesat.gyro()
+        starting_mag = cubesat.magnetic()
+        time.sleep(wait_time)
+        ending_mag = cubesat.magnetic()
+        # calculate the change in the total magnetometer reading
+        mag = (sqrt(ending_mag[0] ** 2 + ending_mag[1] ** 2 +
+               ending_mag[2] ** 2) - sqrt(starting_mag[0] ** 2 +
+               starting_mag[1] ** 2 + starting_mag[2] ** 2))
+        print("Data Collection Complete")
+        return acc, gyro, mag
+
+
+def imu_test(result_dict, test_type):
+    """
+    All automation happens in this function
+    Process user test results and update the result dictionary accordingly
+    """
+    if test_type == "Still":
+        # record acceleration
+        acc, gyro, mag = user_test(test_type)
+        result_val_string_acc = ("Testing IMU when cubesat is still. Acc: " +
+                                 str(acc))
+        result_val_string_gyro = ("Testing IMU when cubesat is still. Gyro: " +
+                                  str(gyro))
+
+        # if total acceleration ~= 9.8 m/s^2 and gyro ~= 0 deg/s, true
+        result_val_bool_acc = abs(
+            9.8 - sqrt(acc[0] ** 2 + acc[1] ** 2 + acc[2] ** 2)) < 0.2
+        result_val_bool_gyro = sqrt(
+            gyro[0] ** 2 + gyro[1] ** 2 + gyro[2] ** 2) < 0.2
+        result_dict['IMU_AccStill'] = (
+            result_val_string_acc, result_val_bool_acc)
+        result_dict['IMU_GyroStill'] = (
+            result_val_string_gyro, result_val_bool_gyro)
+
+    elif test_type == "Moving":
+        # record acceleration
+        acc, gyro, mag = user_test(test_type)
+        result_val_string = ("Testing IMU when cubesat is being moved. Acc: " +
+                             str(acc))
+
+        # if absolute value of each acceleration component is more than 0, true
+        result_val_bool = (abs(acc[0]) > 0 and abs(acc[1]) > 0 and
```

Comment:
This will always be true, there is some nosie.

Reply:
I checked for norm(acc) >= 0.2 instead; this was the threshold I used in the last few tests, so it seems to make the most sense.